#81 Feature/signup save step1

ebPark9511 opened the pull request:
#81 화면, 화면 데이터 저장

ebPark9511 committed:
✨ Add Signup's navigation

ebPark9511 committed:
✅ Add test code about Signup's navigation

Jae-eun commented:
테스트 코드까지 넣으셨군요! 멋집니다~

ebPark9511 committed:
✨ Save Signup enum in Userdefault

ebPark9511 committed:
✨ Save Signup step in UserDefault

ebPark9511 committed:
♻️ Modify class name

ebPark9511 committed:
✨ Add viewData maintain

Commit:
♻️ Modify protocol name

ebPark9511 commented:
화면 데이터 저장작업으로 인해 준영님 재은님 소스를 좀 건드렸는데요!
이상한 부분 있으면 편하게 코멘트 부탁드립니다.

junyng commented on Soyeon/Common/UserDefault/SoyeonUserDefaultKey.swift:
```diff
@@ -0,0 +1,16 @@
+import Foundation
+
+/// Soyeon 서비스에서 사용될 UserDefualt Key 모음.
+enum SoyeonUserDefaultKey: String {
+    
+    // 회원가입
+    case saveSignUpLocation
+}
```
서비스 키를 연관 타입으로 정리해서 좋은것 같아요 👍

junyng commented on SoyeonTests/SignupUserDefault.swift:
```diff
@@ -0,0 +1,34 @@
+//  Copyright © 2021 ludus. All rights reserved.
+//
+
+import XCTest
+@testable import Soyeon
+ 
+class SignupUserDefault: XCTestCase {
+    
+    override func setUpWithError() throws {
+        let saveTarget = Signup.step1(.agreement).path
+        
+        UserDefaults.setValue(saveTarget, forKey: .saveSignUpLocation)
+        
+    }
+  
+    func testsUserDefulat() {
+        if let result = UserDefaults.object(forKey: .saveSignUpLocation) as? String,
+           let saveSignupLocation: Signup = Signup.stringToInit(value: result) {
+            
+            let navigation = saveSignupLocation.loaded
+            
+            let testResult = navigation.viewControllers[0] is LoginViewController &&
+                             navigation.viewControllers[1] is AgreementViewController
+             
+             XCTAssertTrue(testResult)
+        }
+    }
+}
+  
```
테스트 코드 작성하신 것 멋지네용👍

junyng commented on Soyeon/Common/UserDefault/SoyeonUserDefault.swift:
```diff
@@ -0,0 +1,32 @@
+
+import Foundation
+
+protocol SoyeonUserDefault {
+    static func remove(forKey: SoyeonUserDefaultKey) 
+    static func setValue(_ value: Any, forKey: SoyeonUserDefaultKey)
+    static func object(forKey: SoyeonUserDefaultKey) -> Any?
+}
+
+extension SoyeonUserDefault {
+    static func remove(forKey: SoyeonUserDefaultKey) {
+        UserDefaults.standard.removeObject(forKey: forKey.rawValue)
+    }
+    
+    static func setValue(_ value: Any, forKey: SoyeonUserDefaultKey) {
+        UserDefaults.standard.setValue(value, forKey: forKey.rawValue)
+    }
+     
+    static func object(forKey: SoyeonUserDefaultKey) -> Any? {
+        return UserDefaults.standard.object(forKey: forKey.rawValue)
+    }
+}
+
+extension UserDefaults: SoyeonUserDefault {
+}
```
`UserDefaults` 타입에 확장 메서드를 추가하셨네요 👍

junyng commented on Soyeon/Intro/StartSplash/StartSplashViewController.swift:
```diff
@@ -50,28 +50,45 @@ final class StartSplashViewController: UIViewController {
+        dictionary.keys.forEach { key in
+            UserDefaults.standard.removeObject(forKey: key)
+        }
+        
```
지금은 회원가입을 위한 메서드로 사용되는 것 같아 보여요
모든 키에 대한 `UserDefaults` 정보가 사라질 것 같은데 
회원가입 키에 대한 정보만 재설정되어야 할 것 같아요~!

junyng commented on Soyeon/Signup/LoadSignupView.swift:
```diff
@@ -0,0 +1,52 @@
+        let allCases = EnumType.allCases
+         
+        if let firstIndex = allCases.firstIndex(of: until) {
+            let untilViewCntorllerNames = allCases[...firstIndex].map { $0.viewControllerName }
```
Cntorller 오타가 있는거 같아요!

junyng commented on Soyeon/Signup/Step1/Login/Configurator/LoginConfigurator.swift:
```diff
@@ -22,5 +21,7 @@ final class LoginConfigurator {
         viewController.router = router
+        
+        self.step = .step1(.login)
     }
```
데이터 저장에 대한 비즈니스 로직은 `Interactor`에서 수행되어야 할 것 같아요!

ebPark9511 replied:
준영님 지적 감사합니다! 

회원가입 VIewController는 모두 동일한 일(현재화면의 저장, 현재화면의 데이터 저장)을 하고
VIP패턴을 사용하지 않는 경우가 있기 때문에 
Interator에 두는 것보단 UIViewController를 서브클래싱 하는것이 났다는 판단이 들었습니다.

SignupStepSavable, LoadSignupViewData를 채택하는 SignupStepViewController를 만들었습니다!

보시고 의견 있으시면 코멘트 부탁드립니다!

ebPark9511 replied:
준영님 지적 감사합니다! 

회원가입 VIewController는 모두 동일한 일(현재화면의 저장, 현재화면의 데이터 저장)을 하고
VIP패턴을 사용하지 않는 경우가 있기 때문에 
Interator에 두는 것보단 UIViewController를 서브클래싱 하는것이 났다는 판단이 들었습니다.

SignupStepSavable, LoadSignupViewData를 채택하는 SignupStepViewController를 만들었습니다!

보시고 의견 있으시면 코멘트 부탁드립니다!

ebPark9511 replied:
준영님 지적 감사합니다! 

회원가입 VIewController는 모두 동일한 일(현재화면의 저장, 현재화면의 데이터 저장)을 하고
VIP패턴을 사용하지 않는 경우가 있기 때문에 
Interator에 두는 것보단 UIViewController를 서브클래싱 하는것이 났다는 판단이 들었습니다.

SignupStepSavable, LoadSignupViewData를 채택하는 SignupStepViewController를 만들었습니다!

보시고 의견 있으시면 코멘트 부탁드립니다!

ebPark9511 replied:
준영님 지적 감사합니다! 

회원가입 VIewController는 모두 동일한 일(현재화면의 저장, 현재화면의 데이터 저장)을 하고
VIP패턴을 사용하지 않는 경우가 있기 때문에 
Interator에 두는 것보단 UIViewController를 서브클래싱 하는것이 났다는 판단이 들었습니다.

SignupStepSavable, LoadSignupViewData를 채택하는 SignupStepViewController를 만들었습니다!

보시고 의견 있으시면 코멘트 부탁드립니다!

junyng replied:
아 그렇군요 ! 이해하였습니다~
인터렉터가 없는것을 확인못했네요 😅

제 의견을 다시 정리하자면...

`SignupStepSavable`.`step` 프로퍼티의 `getter setter`에서 저장 로직이 동작하고 있어서
(회원 가입을 진행중인 path 저장 동작)을 함수로 정의하면 의도가 명확해질것 같다고 생각했어요~!


```
extension SignupStepSavable {
     var step: Signup? {
         get { return nil }
         set {
             if let newValue = newValue {
                 UserDefaults.setValue(newValue.path,
                                       forKey: .saveSignUpLocation)
             }
         }
     }
```

ebPark9511 replied:
맞는 말씀이십니다 ㅎㅎㅎ 반영 완료 했습니다.
의견 감사합니다!!

junyng reviewed (commented):
작업 많으셨을텐데 ㅜㅜ 고생많으셨습니다~!
리뷰 몇개 남겨보았습니다.!

ebPark9511 committed:
🩹 Code review was reflected.

- modify typo
- SignupStepSavable, LoadSignupViewData -> class SignupStepViewController

ebPark9511 committed:
🩹 Code review was reflected.
